API/ev3_sensors/ev3_gyro.c

```diff
@@ -0,0 +1,90 @@
+#include "ev3_inputs/ev3_input_uart.h"
+#include "ev3_sensors/ev3_sensors.h"
+#include "contrib/lms2012/ev3_basictypes.h"
+#include "include/ev3_command.h"
+#include "ev3_gyro.h"
+#include "utility.h"
+
+#define EV3_GYRO_SENSOR_TYPE                32
+#define EV3_GYRO_SENSOR_DEFAULT_MODE        EV3_GYRO_SENSOR_ANGLE_AND_RATE_MODE
+
+int ev3GyroSoftwareResetOffset[] = {0, 0, 0, 0};
+
+SensorHandler * EV3Gyro = &(SensorHandler){
+    .Init = initEV3GyroSensor,
+    .Exit = exitEV3GyroSensor,
+    .currentSensorMode = {NONE_MODE, NONE_MODE, NONE_MODE, NONE_MODE}
+};
+
+bool initEV3GyroSensor (int port) {
+    setEV3GyroSensorMode(port, EV3_GYRO_SENSOR_DEFAULT_MODE);
+    return true;
+}
+
+
+int ReadEV3GyroSensor(int port, EV3GyroReadingMode mode) {
+    setEV3GyroSensorMode(port, EV3_GYRO_SENSOR_DEFAULT_MODE);
+
+    uint64_t angleAndRate;
+    readFromUART(port, (DATA8 *) &angleAndRate, 4);
+
+    if (mode == EV3GyroAngle) {
+        int angle = getAngleFromAngleAndRate(angleAndRate);
+        return angle - ev3GyroSoftwareResetOffset[port];
+    } else {
+        return getRateFromAngleAndRate(angleAndRate);
+    }
+}
+
+int getAngleFromAngleAndRate (uint64_t angleAndRate) {
+    int value = angleAndRate & 0xFFFF;
+    return handleEV3GyroNegativeValue(value);
+}
+
+int getRateFromAngleAndRate (uint64_t angleAndRate) {
+    int value = (angleAndRate >> 16) & 0xFFFF;
+    return handleEV3GyroNegativeValue(value);
+}
+
+int handleEV3GyroNegativeValue(int value) {
+    if(value & 0x8000) {
+        /**
+         * - the if checks for the sign-bit
+         * - the and-mask removes the sign bit - that gives us how much bigger is
+         *   the actual value compared to the most negative value
+         * - the subtraction just shifts the result so that it matches the
+         *   previous value. In this case, a +1 is also added, presumably to
+         *   prevent the confusing situation that the most negative value does
+         *   not have a positive counterpart.
+         *
+         * Explanation provided by @JakubVanek, see https://github.com/c4ev3/ev3-api/issues/9
+         */
+        value = ((value & 0x7FFF) - 0x7FFF);
+    }
+    return value;
+}
+
+void ResetEV3GyroSensor(int port) {
```

Comment:
I have realized this might not work.
 - N2 and N3 gyros cannot switch out of GYRO-GnA mode (which is the default here) according to [1]. However maybe the stock kernel works this around and it resets the sensor automatically on timeout (but I have not tested it).
 - N4 and newer gyros will not recalibrate on mode switch according to [2].

Solving this may involve the direct commands that can be sent to the sensor [3]. However sadly the N2 and N3 sensors have botched GYRO-GnA so even the direct commands cannot recalibrate it, so perhaps relying on the full reset is the proper way of doing it.

Sources:
1. http://docs.ev3dev.org/projects/lego-linux-drivers/en/ev3dev-buster/sensor_data.html#lego-ev3-gyro-mode3
2. http://ev3lessons.com/en/ProgrammingLessons/advanced/GyroRevisited.pdf
3. https://github.com/ev3dev/ev3dev/issues/1387

Reply:
There's always the option of powering off the sensor completely (perhaps the DCM pin interface could do this somehow) and letting it then start from scratch. Alternatively, the UART layer could somehow be instructed to let the sensor time out and reset on its own. Apart from that, the options I know of are the following:
* GYRO-ANGLE
  - N2, N3
    - mode switch should work well
    - sending `0x11` will reset the offset (but not recalibrate)
    - sending `0x88` will reset the offset and recalibrate
  - N4+
    - not tested, but modeswitch should not work, while the commands should still work.
* GYRO-RATE
  - N2, N3
    - mode switch should work well
    - sending `0x11` will recalibrate the sensor
    - sending `0x88` will recalibrate the sensor too
  - N4+
    - not tested, but modeswitch should not work, while the commands should still work.
* GYRO-GnA
  - N2, N3
    - not known; perhaps the timeout will be triggered automatically
    - according to [2] the stock lms2012 VM could make the newer sensors time out from other modes by reading it as an IR sensor; while it's an ugly hack, it might work
  - N4+
    - not tested, but if it behaves like GYRO-ANGLE and GYRO-RATE combined on N3, then
      - `0x11` should reset the offset of the angle
      - `0x88` should reset the offset and recalibrate both the angle and rate
* GYRO-FAS is not important now
* TILT-ANGLE is not important now
* TILT-RATE is not important now

Reply:
Hmm, it seems that the sensors cannot be powered down on a per-port basis (instead the whole external 5V rail has to be turned off: https://github.com/ev3dev/ev3dev/issues/1206#issuecomment-446390876).

Reply:
On the other hand, I don't think that this is a blocker, as the current API doesn't do better here.

Reply:
I have found out that my summary above is wrong, see https://github.com/ev3dev/ev3dev/issues/1387#issuecomment-596838638. I will update the comment.